File: Cubixcraft_v0.0.3-alpha_2021.05.18_java_src_IntelliJidea/src/game/cubixcraft/utils/Font.java
package game.cubixcraft.utils;

import game.cubixcraft.CubixCraft;

import java.nio.charset.Charset;

import static org.lwjgl.opengl.GL11.*;

public class Font {
    private int textureId;
    private final float fontSize = 16.0f;
    private final float widthCoef = 1.4f;
    private final float[] charWidths;

    public Font(String fileName) {
        /*
         * Загружает текстурный шрифт.
         * Текстура должна быть в формате PNG 32 bits с альфа-каналом
         * и содержать символы таблицы ASCII, 16 штук по ширине и 16 по высоте.
         */
        System.out.println("Loading font...");

        textureId = Textures.loadTexture(fileName);
        charWidths = new float[256];
        for (int i = 0; i < 256; i++) {
            charWidths[i] = fontSize / widthCoef;
        }
        charWidths['l'] = fontSize / 2.0f / widthCoef;
    }

    public void start() {
        glMatrixMode(GL_PROJECTION);
        glLoadIdentity();
        glOrtho(0.0, CubixCraft.getWindowWidth(), CubixCraft.getWindowHeight(), 0.0, -1.0, 1.0);
        glMatrixMode(GL_MODELVIEW);
        glLoadIdentity();

        glEnable(GL_ALPHA_TEST);
        glAlphaFunc(GL_GREATER, 0.0f);
        glDisable(GL_DEPTH_TEST);

        glBindTexture(GL_TEXTURE_2D, textureId);
    }

   public void end() {
        glBindTexture(GL_TEXTURE_2D, 0);
        glDisable(GL_ALPHA_TEST);
        glEnable(GL_DEPTH_TEST);
    }

    public void drawString(String textString, float xPos, float yPos, int color) {
        float r = (color >> 16 & 0xFF) / 255.0f;
        float g = (color >>  8 & 0xFF) / 255.0f;
        float b = (color       & 0xFF) / 255.0f;
        glColor3f(r, g, b);

        glLoadIdentity();

        byte[] chars = textString.getBytes(Charset.forName("cp1251"));
        final int glyphCount = 16;
        final float glyphSize = 1.0f / (float)glyphCount;
        float xOffset = 0.0f;

        glEnable(GL_TEXTURE_2D);

        glBegin(GL_QUADS);
        for (int i = 0; i < chars.length; i++) {
            char charId = (char)(chars[i] - 32); // or int charId = (chars[i] < 0 ? chars[i] + 256 : chars[i]) - 32;
            float texPosX = (charId % glyphCount) * glyphSize;
            float texPosY = (float)(charId / glyphCount) / (float)glyphCount;
            float u1 = texPosX + glyphSize;
            float v1 = texPosY + glyphSize;

            glTexCoord2f(texPosX, texPosY);
            glVertex2f(xPos + xOffset, yPos);
            glTexCoord2f(u1, texPosY);
            glVertex2f(xPos + xOffset + fontSize, yPos);
            glTexCoord2f(u1, v1);
            glVertex2f(xPos + xOffset + fontSize, yPos + fontSize);
            glTexCoord2f(texPosX, v1);
            glVertex2f(xPos + xOffset, yPos + fontSize);

            xOffset += charWidths[chars[i]];
        }
        glEnd();

        glDisable(GL_TEXTURE_2D);
    }

    public void drawBkg(float xPos, float yPos, float width, int color) {
        float r = (color >> 16 & 0xFF) / 255.0f;
        float g = (color >> 8  & 0xFF) / 255.0f;
        float b = (color       & 0xFF) / 255.0f;
        glColor4f(r, g, b, 0.2f);

        glDisable(GL_TEXTURE_2D);
        glEnable(GL_BLEND);
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);

        glBegin(GL_QUADS);
        glVertex2f(xPos, yPos);
        glVertex2f(xPos + width, yPos);
        glVertex2f(xPos + width, yPos + fontSize + 2.0f);
        glVertex2f(xPos, yPos + fontSize + 2.0f);
        glEnd();

        glDisable(GL_BLEND);
    }

    public float getStringWidth(String textString) {
        char[] chars = textString.toCharArray();
        float n = 0.0f;
        for (int i = 0; i < chars.length; i++) {
            n += charWidths[chars[i]];
        }
        return n;
    }

    public int getTextureId() {
        return textureId;
    }

    public void destroy() {
        if (textureId > 0) {
            glDeleteTextures(textureId);
            textureId = 0;
        }
    }

}

File: Cubixcraft_v0.0.3-alpha_2021.05.18_src_VisualC++2019/Cubixcraft/font.cpp
#include "font.h"
#include "main.h"
#include "textures.h"

GLuint fontTex = 0;
GLuint glyphLists = 0;

/*
 * BuildTexturedFont()
 * Загружает текстурный шрифт.
 * Текстура должна быть в формате PNG 32 bits с альфа-каналом
 * и содержать символы таблицы ASCII, 16 штук по ширине и 16 по высоте.
 */
GLuint BuildTexturedFont(std::string textureFileName) 
{
    fontTex = LoadTexture(textureFileName);
    if (fontTex)
    {
        const int symbolsInRow = 16;
        float glyphSize = 1.0f / (float)symbolsInRow;
        const float shiftX = fontSize / symbolWidthCoef;

        //создаём отдельные листы для каждой буквы
        glyphLists = glGenLists(256);

        for (int i = 0; i < 256; i++)
        {
            float glyphPosX = (i % symbolsInRow) / (float)symbolsInRow;
            float glyphPosY = (i / symbolsInRow) / (float)symbolsInRow;
            float u1 = glyphPosX + glyphSize;
            float v1 = 1.0f - glyphPosY;

            glNewList(glyphLists + i, GL_COMPILE);

            glBegin(GL_QUADS);

            glTexCoord2f(glyphPosX, v1 - glyphSize);
            glVertex2f(0.0f, 0.0f);
            
            glTexCoord2f(u1, v1 - glyphSize);
            glVertex2f(fontSize, 0.0f);

            glTexCoord2f(u1, v1);
            glVertex2f(fontSize, fontSize);

            glTexCoord2f(glyphPosX, v1);
            glVertex2f(0.0f, fontSize);

            glEnd();

            glTranslatef(shiftX, 0.0f, 0.0f);

            glEndList();
        }
        std::cout << "Loaded font: " << textureFileName << std::endl;
    }
    return fontTex;
}

/*
 * KillFont()
 * Выгружает текстуру шрифта и освобождает память.
 */
void KillFont() 
{
    if (glyphLists) 
    {
        glDeleteLists(glyphLists, 256);
        glyphLists = 0;
    }
    if (fontTex) 
    {
        glDeleteTextures(1, &fontTex);
        fontTex = 0;
    }
}

/*
 * glTextStart()
 * Устанавливает необходимые параметры для вывода текста
 * и переключает OpenGL в 2D-режим рисования (ортогональная проекция).
 */
void glTextStart() 
{
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();

    glOrtho(0.0, (double)viewportWidth, 0.0, (double)viewportHeight, -1.0, 1.0);

    glMatrixMode(GL_MODELVIEW);
    glLoadIdentity();
    
    glEnable(GL_ALPHA_TEST);
    glAlphaFunc(GL_GREATER, 0.0f);
    glDisable(GL_DEPTH_TEST);

    glBindTexture(GL_TEXTURE_2D, fontTex);
}

/*
 * glTextEnd()
 * Завершает вывод текста.
 */
void glTextEnd() 
{
    glBindTexture(GL_TEXTURE_2D, 0);

    glEnable(GL_DEPTH_TEST);
    glDisable(GL_ALPHA_TEST);
}

/*
 * glTextOut()
 * Выводит текст на экран.
 */
void glTextOut(const char* sText, float xPos, float yPos, int color)
{
    float r = (color >> 16 & 0xFF) / 255.0f;
    float g = (color >> 8  & 0xFF) / 255.0f;
    float b = (color       & 0xFF) / 255.0f;
    glColor3f(r, g, b);

    glLoadIdentity();
    glTranslatef(xPos, (float)viewportHeight - fontSize - yPos - 4.0f, 0.0f);

    glListBase(glyphLists - 32);
    glEnable(GL_TEXTURE_2D);
    glCallLists(strlen(sText), GL_UNSIGNED_BYTE, sText);
    glDisable(GL_TEXTURE_2D);
}

/*
 * DrawTextBkg()
 * Рисует полупрозрачный прямоугольник.
 */
void DrawTextBkg(float xPos, float yPos, float xWidth, int color)
{
    float r = (color >> 16 & 0xFF) / 255.0f;
    float g = (color >>  8 & 0xFF) / 255.0f;
    float b = (color       & 0xFF) / 255.0f;
    glColor4f(r, g, b, 0.2f);

    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    
    glLoadIdentity();
    
    glBegin(GL_QUADS);
    glVertex2f(xPos,          (float)viewportHeight - yPos - 20.0f);
    glVertex2f(xPos + xWidth, (float)viewportHeight - yPos - 20.0f);
    glVertex2f(xPos + xWidth, (float)viewportHeight - yPos        );
    glVertex2f(xPos,          (float)viewportHeight - yPos        );
    glEnd();

    glDisable(GL_BLEND);
}

float GetStringWidth(std::string textString)
{
    return (textString.length() * fontSize) / symbolWidthCoef;
}

std::string FormatString(char* sText, ...)
{
    char text[256];
    va_list vl;
    va_start(vl, sText);
    vsprintf_s(text, sText, vl);
    va_end(vl);
    return text;
}
